hepcrawl/spiders/elsevier.py

```diff
@@ -0,0 +1,235 @@
+                urls_for_packages[file_data[0].text] = link
+        return urls_for_packages
+
+    def _get_all_new_packages(self, elsevier_metadata):
+        """
+        Checks which packages from elsevier batch feed are not in the s3 bucket yet
+        Returns:
+            dict(name: url): dict of zip packages names and urls
+
+        """
+        urls_for_packages = self._get_package_urls_from_elsevier(elsevier_metadata)
+        bucket_data = self._get_keys_names_from_bucket()
+        packages_not_in_bucket = {
+            name: urls_for_packages[name]
+            for name in set(urls_for_packages.keys()) - bucket_data
+        }
+        self.new_packages = set(packages_not_in_bucket.keys())
+        return packages_not_in_bucket
+
+    def start_requests(self):
+        elsevier_batch_download_url = self.elsevier_consyn_url
+        yield Request(
+            elsevier_batch_download_url, callback=self.get_packages_from_elsevier
+        )
+
+    def get_packages_from_elsevier(self, response):
+        """
+        Parse batch feed file from elsevier and downloads new zip packages from Elsevier server.
+        """
+        elsevier_metadata = response.body
+        for name, url in self._get_all_new_packages(elsevier_metadata).items():
+            if name.lower().endswith("zip"):
+                yield Request(
+                    url,
+                    callback=self.populate_s3_bucket_with_elsevier_packages,
+                    meta={"name": name},
+                )
+
+    def populate_s3_bucket_with_elsevier_packages(self, response):
+        """
+        Uploads to s3 bucket new zip packages.
+        """
+        name = response.meta["name"]
+        url = self.create_presigned_url(
+            method="put_object", bucket=self.packages_bucket_name, file=name
+        )
+        yield Request(
+            url,
+            method="PUT",
+            body=response.body,
+            meta={"name": name},
+            callback=self.download_zip_packages,
+        )
+
+    @staticmethod
+    def _get_doi_for_xml_file(xml_file):
+        parser = ElsevierParser(xml_file)
+        doi = parser.get_identifier()
+        return doi
+
+    def download_zip_packages(self, response):
+        """
+        Downloads zip packages from s3 and passes to unziping function.
+        """
+        get_url = self.create_presigned_url(
+            bucket=self.packages_bucket_name,
+            method="get_object",
+            file=response.meta["name"],
+        )
+        yield Request(
+            get_url,
+            callback=self.unzip_zip_package_to_s3,
+            meta={"name": response.meta["name"]},
+        )
+
+    def unzip_zip_package_to_s3(self, response):
+        """
+        Extracts the files from zip folders downloaded from elsevier and
+        uploads them with a correct name (article doi) to the correct s3 bucket.
+        """
+        tempdir = tempfile.mkdtemp()
```

Comment:
you should use https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory (or a backport if not available on Python 2.7) which can be used as a context manager. Your code has a bug, it's not cleaning up if there are exceptions.